Refresh.GameServer/Endpoints/Game/Playlists/PlaylistEndpoints.cs

```diff
@@ -0,0 +1,338 @@
+using Bunkum.Core.Endpoints;
+using Bunkum.Core.Endpoints.Debugging;
+using Bunkum.Core.Responses;
+using Bunkum.Listener.Protocol;
+using Bunkum.Protocols.Http;
+using Refresh.GameServer.Database;
+using Refresh.GameServer.Extensions;
+using Refresh.GameServer.Types.Data;
+using Refresh.GameServer.Types.Levels;
+using Refresh.GameServer.Types.Lists;
+using Refresh.GameServer.Types.Playlists;
+using Refresh.GameServer.Types.Roles;
+using Refresh.GameServer.Types.UserData;
+
+namespace Refresh.GameServer.Endpoints.Game.Playlists;
+
+public class PlaylistEndpoints : EndpointGroup
+{
+    // Creates a playlist, with an optional parent ID
+    [GameEndpoint("createPlaylist", HttpMethods.Post, ContentType.Xml)]
+    [RequireEmailVerified]
+    public Response CreatePlaylist(RequestContext context, DataContext dataContext, SerializedPlaylist body)
+    {
+        GameUser user = dataContext.User!;
+        
+        GamePlaylist? parent = null;
+        // If the parent ID is specified, try to parse that out
+        if (int.TryParse(context.QueryString["parent_id"], out int parentId))
+        {
+            parent = dataContext.Database.GetPlaylistById(parentId);
+
+            // Dont try to parent to a non-existent parent playlist
+            if (parent == null)
+                return BadRequest;
+            
+            // Dont let you create a sub-playlist of someone else's playlist
+            if (user.UserId != parent.Creator.UserId)
+                return Unauthorized;
+
+            // If the user has no root playlist, but they are trying to create a sub-playlist, something has gone wrong.
+            if (user.RootPlaylist == null)
+                return BadRequest;
+        }
+
+        // Create the playlist, marking it as the root playlist if the user does not have one set already
+        GamePlaylist playlist = dataContext.Database.CreatePlaylist(user, body, user.RootPlaylist == null);
+
+        // If there is a parent, add the new playlist to the parent
+        if (parent != null) 
+            dataContext.Database.AddPlaylistToPlaylist(playlist, parent);
+
+        // If this new playlist is the root playlist, mark the user's root playlist as it
+        if (playlist.RootPlaylist)
+            dataContext.Database.SetUserRootPlaylist(user, playlist);
+        
+        // Create the new playlist, returning the data
+        return new Response(SerializedPlaylist.FromOld(playlist, dataContext), ContentType.Xml);
+    }
+
+    // Gets the slots contained within a playlist
+    [GameEndpoint("playlist/{id}", HttpMethods.Get, ContentType.Xml)]
+    [NullStatusCode(NotFound)]
+    [MinimumRole(GameUserRole.Restricted)]
+    public SerializedMinimalLevelList? GetPlaylistSlots(RequestContext context, DataContext dataContext, int id)
+    {
+        GamePlaylist? playlist = dataContext.Database.GetPlaylistById(id);
+
+        // Handle an invalid playlist ID
+        if (playlist == null)
+            return null;
```

Comment:
Don't think this needs a comment.
```suggestion
        if (playlist == null)
            return null;
```